hedera-node/test-clients/scripts/diff-testing/main.py

```diff
@@ -0,0 +1,233 @@
+
+import copy
+
+from bloblist import Blob, BlobList
+from cli_arguments import Args, a
+from gcp_access import *
+from globals import Streams, g
+from utils import *
+
+
+# Configuration setup functions
+
+def set_derived_configuration():
+    """Set configuration settings that are derived, by some formula, from other configuration settings"""
+
+    g.interval_start_time_with_T = a.interval_start_time.isoformat('T', 'seconds')
+    g.interval_end_time_with_T = a.interval_end_time.isoformat('T', 'seconds')
+
+    g.interval_start_time_as_filename_component = g.interval_start_time_with_T.replace(':', '_')
+    g.interval_end_time_as_filename_component = g.interval_end_time_with_T.replace(':', '_')
+    g.interval_common_prefix = os.path.commonprefix(
+        [g.interval_start_time_as_filename_component, g.interval_end_time_as_filename_component])
+
+    if not g.interval_common_prefix:
+        print(f"*** Possible problem: interval_comon_prefix has 0-length; will match all timestamps, no filtering")
+
+
+def set_fixed_configuration():
+    """Set configuration settings that are so fixed you can just put them in the code"""
+    g.n_nodes = C.N_NODES
+
+
+def set_default_paths_and_patterns():
+    """Set default paths and path patterns (path patterns: GCP patterns for Hedera stream files)"""
+    g.gcp_project = C.GCP_PROJECT
+    g.network = C.NETWORK
+    g.gcp_export_bucket = C.GCP_EXPORT_BUCKET
+
+    # each pattern has two arguments in this order: ( node_number, timestamp_common_prefix )
+    g.stream_patterns = {Streams.EVENTS: 'eventsStreams/events_0.0.{}/{}*.evts',
+                         Streams.RECORDS: 'recordstreams/record0.0.{}/{}*.rcd.gz',
+                         Streams.SIDECARS: 'recordstreams/record0.0.{}/sidecar/{}*.rcd.gz'}
+
+    g.match_node_number_re = re.compile(r'_?0[.]0[.][0-9]{1,2}')
+
+
+def read_mainnet_hostnames():
+    """Read the map from node# to hostnames, from a file"""
+    g.mainnet_hostnames_path = f"{a.local_storage_rootdir}/{a.hostnamemap_filename}"
+    f = open(g.mainnet_hostnames_path, 'r')
+    lines = f.readlines()
+    lines = [line.removeprefix('node') for line in lines]
+    lines = [line.split() for line in lines]
+    lines = [(int(line[0]), line[1]) for line in lines]
+    g.mainnet_hostnames = dict(lines)
+    # [print(k, ':', v) for k, v in MAINNET_HOSTNAMES.items()]
+
+
+def get_ready_to_go():
+    """Get all configuration settings, or provide defaults"""
+    set_derived_configuration()
+    set_fixed_configuration()
+    set_default_paths_and_patterns()
+    read_mainnet_hostnames()
+
+
+# node# <=> hostname mapping functions
+
+def get_name_of_node(node_number):
+    return g.mainnet_hostnames[node_number]
+
+
+def get_number_of_node(node_name):
+    return next(k for k, v in g.mainnet_hostnames.items() if v == node_name)
+
+
+# Handle lists of available blobs in gcp
+
+def get_blobs_for_stream(bucket, stream):
```

Comment:
I like the idea of having a list of blobs that we can check before downloading, well done :)

Reply:
TY. It also lets you do two things: Make sure you have a full set - suppose a node is down (on it's own, not part of a system upgrade) and is missing files within the interval.  You rerun specifying a _different_ node and it'll "fill in the gap".  Second, it then keeps track of successful downloads - you keep rerunning the download until all files are present.  Otherwise transient gcp/network problems will give you missing files.

Reply:
Yes, both of which I've run into - ty!